### ISSUE TITLE
Tanh produces incorrect results

### ISSUE THREAD

**Issue by neslib** [neslib/Neslib.MultiPrecision#3]
As reported by @Andreas113

Hi Erik,
In the meantime, I made several more numerical tests of the mathematical routines and I came across the following error:

```Delphi
Function: Hyperbolic tangent: Tanh(..)
```

Numerical Example:

Tanh(Y/X):
----------
X            =  3.42897556698888888812345
Y            = +1.63128799224455996622337700123
MPA_Float    = 0.442822429892976193018935445355621315339799605925608439366306296677512: Exact!
DoubleDouble = 0.4428224298929761930189354453556
QuadDouble   = 0.44282242989297619301893544535562131533979960592560843936630630

Conclusion:
DobleDouble: -----> OK!
QuadDouble : -----> OK!

Tanh(Y/X):
----------
X            = -3.5
Y            =  1.0E-1
MPA_Float    = -0.028563656570828037865016530810004960582783350513412052472238074033367: Exact!
DoubleDouble = -0.2781854903257024404718000872415 -----> WRONG!
QuadDouble   = -0.27818549032570244047180008724146610677969449455154405945034634 -----> WRONG!

Conclusion:
DobleDouble: -----> WRONG!
QuadDouble : -----> WRONG!
On Delphi XE5 still only DoubleDouble runs, on Delphi 10.3 DobleDouble and QuadDouble.
The wrong result comes out - with this number combination - with both Delphi versions!
Even my old calculator (HP 41 CX) can calculate this value to 10 digits correctly...
So, there seems to be something wrong with the C routine. Or is it possibly due to the data transfer from Delphi to C?

I’ will be keeping testing!

Regards,
Andreas

**Comment by Andreas113** [comment]
Erik,
I made another purposeful observation:
**Function: Division**
Example:
```
Y/X:
----
X            = -3.5
Y            =  0.1
MPA_Float    = -0.028571428571428571428571428571428571428571428571428571428571428571429
QuadDouble   = -0.02857142857142857142857142857142857142857142857142857142857143
DoubleDouble = -0.0285714285714285714285714285714

Conclusion:
DobleDouble: -----> OK!
QuadDouble : -----> OK!

Y/X: Which is perfectly equal, only in exponential notation from above.
----
X            = -3.5
Y            =  1.0E-1
MPA_Float    = -0.028571428571428571428571428571428571428571428571428571428571428571429
QuadDouble   = -0.28571428571428571428571428571428571428571428571428571428571429 --> WRONG!
DoubleDouble = -0.2857142857142857142857142857143 --> WRONG!

Conclusion:
DobleDouble: -----> WRONG!
QuadDouble : -----> WRONG!
```
The Exponent **E-1** is not parsed correctly.

It **all** seems to be due to the parser routine, including the “obfuscated” apparent error of Hyperbolic tangent: Tanh(..)

Andreas

**Comment by neslib** [comment]
This issue is fixed by the fix for issue #4. Not only was there an infinite loop while parsing the exponent; It also didn't parse negative exponents correctly.
I added a unit test for this issue as well.
Thanks for reporting!

Let me know if this solves your issue.
